hbase-server/src/main/java/org/apache/hadoop/hbase/master/balancer/HeterogeneousRegionCountCostFunction.java

```diff
@@ -0,0 +1,288 @@
+ * In order to use this CostFunction, you need to set the following options:
+ * <ul>
+ *     <li>hbase.master.balancer.stochastic.additionalCostFunctions</li>
+ *     <li>hbase.master.balancer.stochastic.heterogeneousRegionCountRulesFile</li>
+ *     <li>hbase.master.balancer.stochastic.heterogeneousRegionCountDefault</li>
+ * </ul>
+ */
+@InterfaceAudience.Private
+public class HeterogeneousRegionCountCostFunction extends StochasticLoadBalancer.CostFunction {
+
+  /**
+   * configuration used for the path where the rule file is stored.
+   */
+  static final String HBASE_MASTER_BALANCER_HETEROGENEOUS_RULES_FILE =
+          "hbase.master.balancer.heterogeneousRegionCountRulesFile";
+  private static final Logger LOG = LoggerFactory.getLogger(
+          HeterogeneousRegionCountCostFunction.class);
+  /**
+   * Default rule to apply when the rule file is not found. Default to 200.
+   */
+  private static final String HBASE_MASTER_BALANCER_HETEROGENEOUS_RULES_DEFAULT =
+          "hbase.master.balancer.heterogeneousRegionCountDefault";
+  /**
+   * Cost for the function. Default to 500, can be changed.
+   */
+  private static final String REGION_COUNT_SKEW_COST_KEY =
+          "hbase.master.balancer.stochastic.heterogeneousRegionCountCost";
+  private static final float DEFAULT_REGION_COUNT_SKEW_COST = 5000;
+  private final String rulesPath;
+
+  /**
+   * Contains the rules, key is the regexp for ServerName, value is the limit
+   */
+  private final Map<Pattern, Integer> limitPerRule;
+
+  /**
+   * This is a cache, used to not go through all the limitPerRule map when searching for limit
+   */
+  private final Map<ServerName, Integer> limitPerRS;
+  private int defaultNumberOfRegions;
+
+  /**
+   * Total capacity of regions for the cluster, based on the online RS and their associated rules
+   */
+  private int totalCapacity = 0;
+
+
+  public HeterogeneousRegionCountCostFunction(final Configuration conf) {
+    super(conf);
+    this.limitPerRS = new HashMap<>();
+    this.limitPerRule = new HashMap<>();
+    this.setMultiplier(conf.getFloat(REGION_COUNT_SKEW_COST_KEY, DEFAULT_REGION_COUNT_SKEW_COST));
+
+    this.rulesPath = conf.get(HBASE_MASTER_BALANCER_HETEROGENEOUS_RULES_FILE);
+
+    this.defaultNumberOfRegions = conf.getInt(
+            HBASE_MASTER_BALANCER_HETEROGENEOUS_RULES_DEFAULT, 200);
+    if (this.defaultNumberOfRegions < 0) {
+      LOG.warn("invalid configuration '" + HBASE_MASTER_BALANCER_HETEROGENEOUS_RULES_DEFAULT
+              +"'. Setting default to 200");
+      this.defaultNumberOfRegions = 200;
+    }
+
+    if (conf.getFloat(
+            StochasticLoadBalancer.RegionCountSkewCostFunction.REGION_COUNT_SKEW_COST_KEY,
+            StochasticLoadBalancer.RegionCountSkewCostFunction.DEFAULT_REGION_COUNT_SKEW_COST) > 0)
+    {
+      LOG.warn("regionCountCost is not set to 0, "
+              + " this will interfere with the HeterogeneousRegionCountCostFunction!");
+    }
```

Comment:
Why is this a problem? Shouldn't we abort this cost function loading if this will cause it to behave improperly?

Reply:
`regionCountCost` is trying to put all RS to the same number of regions. I'm not sure that we should fire a Exception, and skip the init of the cost function